### PR TITLE
[BOOST-4672] implement flexible fee structures for `QuestBudget`

### PR DESCRIPTION
**Pull request by mmackz**
# Management Fee Implementation for QuestBudget

## Overview
This PR implements the management fee feature for the QuestBudget contract, allowing boost creators to earn a fee for managing boosts.

## Key Changes

1. **Management Fee Setting**: 
   - Added `setManagementFee` function to allow the contract owner to set the management fee percentage.
   - Implemented `managementFee` state variable to store the fee percentage.

2. **Boost Creation with Management Fee**:
   - Modified `createERC20Quest` to calculate and reserve the management fee during quest creation.
   - Added `reservedFunds` state variable to track funds reserved for management fees.
   - Updated `available` function to account for reserved funds.

3. **Management Fee Payment**:
   - Implemented `payManagementFee` function to allow boost creators to claim their management fee after completion.
   - Added checks to ensure only the boost creator can claim the fee and only after the remaining rewards have been withdrawn.

4. **Manager Tracking**:
   - Added `questManagers` mapping to associate quest IDs with their creators' addresses.

5. **Events**:
   - Added `ManagementFeeSet` event for logging changes to the management fee percentage.
   - Added `ManagementFeePaid` event for logging management fee payments.

6. **Testing**:
   - Implemented comprehensive test suite for new management fee functionality.
   - Added helper functions and structs to streamline quest creation using common parameters.

## Implementation Details

- The management fee is set in basis points (e.g., 500 = 5%).
- The maximum management fee that can be set is 10000 (100%)
- The fee is calculated based on the actual number of participants rather than the total possible participants.
- Reserved funds are tracked separately to ensure accurate reporting of available funds.
- Safeguards are in place to prevent unauthorized fee claims and to handle edge cases like insufficient funds.

## Security Considerations

- The `setManagementFee` function is restricted to the contract owner.
- The `payManagementFee` function includes checks to prevent unauthorized claims.
- The contract uses OpenZeppelin's `SafeERC20` for secure token transfers.

## Testing

- Setting and updating the management fee
- Creating quests with management fees
- Paying out management fees under various scenarios
- Edge cases such as insufficient funds and unauthorized access attempts

### CODE REVIEW

File: test/QuestBudget.t.sol [hunk 1]
```diff
@@ -368,30 +391,21 @@ contract QuestBudgetTest is Test, TestUtils, IERC1155Receiver {
...
     }
 
-    ///////////////////////////
+    //////////////////////////////////
     // QuestBudget.createERC20Quest //
-    ///////////////////////////
+    //////////////////////////////////
     function testCreateERC20Quest() public {
-        // Define the parameters for the new quest
-        uint32 txHashChainId_ = 1;
```
Comment (mmackz) [inline test/QuestBudget.t.sol:376]: Common parameters for quest creation are moved to helper functions at the bottom of this file `setupQuestData` and `createQuest`. This is to avoid having to repeat the same quest setup on multiple tests.

**Review by topocount — APPROVED** [review]
I don't have ownership permissions on this repo but this looks good to me

File: test/QuestBudget.t.sol [hunk 2]
```diff
@@ -954,10 +1056,248 @@ contract QuestBudgetTest is Test, TestUtils, IERC1155Receiver {
...
+
+        // Get final balance of the quest creator
+        uint256 finalBalance = mockERC20.balanceOf(address(this));
+
+        // Verify the correct amount was transferred
+        assertEq(finalBalance - initialBalance, expectedFeeToPay, "Incorrect management fee paid");
+    }
+
+    function testPayManagementFee_NotAuthorized() public {        
+        vm.prank(address(0xc0ffee));
```
[the remark this answers is not in the document]
Comment (mmackz) [reply test/QuestBudget.t.sol:1219]: Thanks for the info, I did not know about hoax.

**Review by Quazia — APPROVED** [review]
Thank you
